csi_camera.cpp

```diff
@@ -0,0 +1,451 @@
+                video_path = DEFAULT_INPUT_SENSOR;
+            }
+
+            if (attrs->count("debug") == 1) {
+                std::shared_ptr<ProtoType> debug_proto = attrs->at("debug");
+                auto debug_value = debug_proto->proto_value();
+                if (debug_value.has_bool_value()) {
+                    bool debug_bool = static_cast<bool>(debug_value.bool_value());
+                    debug = debug_bool;
+                }
+            }
+        }
+
+        // OVERRIDE
+
+        void reconfigure(Dependencies deps, ResourceConfig cfg) override {
+            if (debug) {
+                std::cout << "Reconfiguring CSI Camera module" << std::endl;
+            }
+
+            // Update attributes
+            auto attrs = cfg.attributes();
+            validate_attrs(attrs);
+
+            // Stop gst pipeline
+            stop_pipeline();
+
+            // Create GST pipeline
+            std::string pipeline_args = create_pipeline();
+            if (debug) {
+                std::cout << "pipeline_args: " << pipeline_args << std::endl;
+            }
+
+            // Start GST pipeline
+            init_csi(pipeline_args);
+        }
+
+        raw_image get_image(std::string mime_type) override {
+            if (debug) {
+                std::cout << "hit get_image. expecting mime_type " << mime_type << std::endl;
+            }
+            raw_image image;
+            image.mime_type = DEFAULT_OUTPUT_MIMETYPE;
+            if (FAKE_CAMERA) {
+                image.bytes = get_test_image();
+                return image;
+            } else {
+                // TODO: handle no bytes retrieved
+                image.bytes = get_csi_image();
+            }
+
+            return image;
+        }
+
+        API dynamic_api() const override {
+            return Camera::static_api();
+        }
+
+        AttributeMap do_command(AttributeMap command) override {
+            std::cerr << "do_command not implemented" << std::endl;
+            return 0;
+        }
+
+        point_cloud get_point_cloud(std::string mime_type) override {
+            std::cerr << "get_point_cloud not implemented" << std::endl;
+            return point_cloud{};
+        }
+        std::vector<GeometryConfig> get_geometries() override {
+            std::cerr << "get_geometries not implemented" << std::endl;
+            return std::vector<GeometryConfig>{};
+        }
+
+        properties get_properties() override {
+            std::cerr << "get_properties not implemented" << std::endl;
+            return properties{};
+        };
+
+        // GST
+
+        void init_csi(std::string pipeline_args) {
+            // Build gst pipeline
+            pipeline = gst_parse_launch(
+                pipeline_args.c_str(),
+                nullptr
+            );
+            if (!pipeline) {
+                std::cerr << "Failed to create the pipeline" << std::endl;
+                std::exit(EXIT_FAILURE);
```

Comment:
If this causes the process to exit, then that will kill the module process, which is a problem if there are multiple instances running within the module. Instead, just this instance should behave as closed. Same for the other EXITs.

Reply:
Interesting, can you explain a little more wrt multiple instances? I thought I wanted to exit the entire process.

Reply:
So any specific module supports one or more particular models of hardware. Here, it supports this specific type of camera. But if someone has two (or more) of these cameras on the same robot, then they'll both be using this same module. In such a case, there would be multiple configured resources in the robot, with different names, but each would be the same model, and would be supported by the same module. AddResource() would get called twice, once for each instance. Closing or removing one camera from the config should never interfere with the other camera. And even if ALL cameras are removed, the module itself should stay running (though pretty much in an idle state) until the module itself is Close()ed as part of robot shutdown, as it's possible someone may add a camera back if they're adjusting configs, and the module itself needs to remaining listening for another AddResource() call in case that happens.

Reply:
Ok this is all making sense - if simply exit it will kill all of the instantiations. 

For example, [here](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/sdk/services/mlmodel/server.cpp#L114C19-L118C24) the ml model service handles exceptions and signals over grpc.

I need to add this status wrapper to all the methods in [components/camera/server.cpp](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/sdk/components/camera/server.cpp) to handle throws and send a failed status.

Reply:
Since this requires an sdk change, I am wondering if i should crete multi-cam support ticket and save for subsequent PR.

Reply:
I think that makes sense.